### ISSUE TITLE
AVX512 CPU kernels result in 5+% slower Transformer training on Cascade Lake CPU

### ISSUE THREAD

**Issue by malfet** [pytorch/pytorch#80252]
### 🐛 Describe the bug

Benchmarking latest nightly using [WLM_Transformers example](https://github.com/pytorch/examples/tree/main/word_language_model)  AVX512 kernels are 18% slower than using AVX2 ones:
```
(py38-cpu-perf-regression) nshulga@compile-st-c524xlarge-1:/fsx/users/nshulga/examples/word_language_model$ ATEN_CPU_CAPABILITY=avx512 python3 main.py --epochs 1 --model Transformer
| epoch   1 |   200/ 2983 batches | lr 20.00 | ms/batch 116.19 | loss 10.18 | ppl 26302.43
| epoch   1 |   400/ 2983 batches | lr 20.00 | ms/batch 114.48 | loss  8.87 | ppl  7147.02
| epoch   1 |   600/ 2983 batches | lr 20.00 | ms/batch 114.27 | loss  8.80 | ppl  6629.19
| epoch   1 |   800/ 2983 batches | lr 20.00 | ms/batch 114.25 | loss  8.72 | ppl  6129.05
| epoch   1 |  1000/ 2983 batches | lr 20.00 | ms/batch 114.11 | loss  8.59 | ppl  5365.68
| epoch   1 |  1200/ 2983 batches | lr 20.00 | ms/batch 114.22 | loss  8.51 | ppl  4983.37
| epoch   1 |  1400/ 2983 batches | lr 20.00 | ms/batch 114.18 | loss  8.39 | ppl  4398.82
| epoch   1 |  1600/ 2983 batches | lr 20.00 | ms/batch 114.15 | loss  8.39 | ppl  4384.90
| epoch   1 |  1800/ 2983 batches | lr 20.00 | ms/batch 118.34 | loss  8.90 | ppl  7346.31
| epoch   1 |  2000/ 2983 batches | lr 20.00 | ms/batch 114.09 | loss  8.44 | ppl  4624.59
| epoch   1 |  2200/ 2983 batches | lr 20.00 | ms/batch 114.31 | loss  8.47 | ppl  4756.44
| epoch   1 |  2400/ 2983 batches | lr 20.00 | ms/batch 116.08 | loss  8.35 | ppl  4221.11
| epoch   1 |  2600/ 2983 batches | lr 20.00 | ms/batch 116.65 | loss  7.94 | ppl  2804.23
| epoch   1 |  2800/ 2983 batches | lr 20.00 | ms/batch 117.27 | loss  7.73 | ppl  2286.98
-----------------------------------------------------------------------------------------
| end of epoch   1 | time: 373.57s | valid loss  7.68 | valid ppl  2172.06
-----------------------------------------------------------------------------------------
=========================================================================================
| End of training | test loss  7.64 | test ppl  2074.23
=========================================================================================
```
```
(py38-cpu-perf-regression) nshulga@compile-st-c524xlarge-1:/fsx/users/nshulga/examples/word_language_model$ ATEN_CPU_CAPABILITY=avx2 python3 main.py --epochs 1 --model Transformer
| epoch   1 |   200/ 2983 batches | lr 20.00 | ms/batch 99.94 | loss 10.23 | ppl 27662.15
| epoch   1 |   400/ 2983 batches | lr 20.00 | ms/batch 99.28 | loss  8.90 | ppl  7320.05
| epoch   1 |   600/ 2983 batches | lr 20.00 | ms/batch 99.09 | loss  8.74 | ppl  6236.06
| epoch   1 |   800/ 2983 batches | lr 20.00 | ms/batch 98.95 | loss  8.66 | ppl  5758.26
| epoch   1 |  1000/ 2983 batches | lr 20.00 | ms/batch 98.71 | loss  8.71 | ppl  6060.82
| epoch   1 |  1200/ 2983 batches | lr 20.00 | ms/batch 98.84 | loss  8.70 | ppl  5988.82
| epoch   1 |  1400/ 2983 batches | lr 20.00 | ms/batch 98.98 | loss  8.45 | ppl  4672.35
| epoch   1 |  1600/ 2983 batches | lr 20.00 | ms/batch 98.90 | loss  8.49 | ppl  4854.61
| epoch   1 |  1800/ 2983 batches | lr 20.00 | ms/batch 98.80 | loss  8.44 | ppl  4621.51
| epoch   1 |  2000/ 2983 batches | lr 20.00 | ms/batch 99.10 | loss  8.52 | ppl  5006.33
| epoch   1 |  2200/ 2983 batches | lr 20.00 | ms/batch 98.87 | loss  8.35 | ppl  4226.65
| epoch   1 |  2400/ 2983 batches | lr 20.00 | ms/batch 98.74 | loss  8.28 | ppl  3956.08
| epoch   1 |  2600/ 2983 batches | lr 20.00 | ms/batch 98.85 | loss  8.49 | ppl  4862.32
| epoch   1 |  2800/ 2983 batches | lr 20.00 | ms/batch 98.68 | loss  8.30 | ppl  4042.19
-----------------------------------------------------------------------------------------
| end of epoch   1 | time: 315.10s | valid loss  7.60 | valid ppl  2003.26
-----------------------------------------------------------------------------------------
=========================================================================================
| End of training | test loss  7.55 | test ppl  1891.44
=========================================================================================
```

### Versions

1.13.0.dev20220610+cpu)

cc @ezyang @gchanan @zou3519 @VitalyFedyunin @ngimel @xuzhao9 @imaginary-person

**Comment by imaginary-person** [comment]
Oops! I hope this wouldn't affect the PyTorch 1.12 release plan.

I'll submit a PR to disable AVX512 dispatch for those kernels that are slower with AVX512, i.e. enable AVX512 dispatch only for those kernels that are faster with AVX512. Perhaps we can also enable/disable AVX512 dispatch according to the platform, as more recent or upcoming platforms might have better AVX512 performance. Regardless, I'll see if there's a way to boost AVX512 performance of existing ATen kernels as well. Memory-bound kernels are more likely to be slower with AVX512, than with AVX2.

**Comment by malfet** [comment]
@imaginary-person for the time being I plan to disable it across the board by default, because at least on any of the workflows we usually benchmarked it did not yield any gains on EC2 C5 machine. Let's continue discussion here on some microbenchmarks one could have run to more easily identify what is going on.

**Comment by malfet** [comment]
By dropping (profile.py)[https://gist.github.com/malfet/0a063a0663230a4cdcd2eb75324e221a] into (https://github.com/pytorch/examples/tree/main/word_language_model) int looks like `aten::_log_softmax ` is the one that regressed the most:
```
(py38-cpu-perf-regression) nshulga@compile-st-c524xlarge-1:/fsx/users/nshulga/examples/word_language_model$ ATEN_CPU_CAPABILITY=avx512 python3 profile.py|head -n10
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                     aten::_log_softmax        35.45%     381.915ms        35.45%     381.915ms      38.191ms            10  
                                               aten::mm        15.67%     168.789ms        15.67%     168.840ms     938.000us           180  
                                            aten::addmm         5.93%      63.872ms        10.39%     111.928ms       1.244ms            90  
                                            aten::fill_         5.76%      62.053ms         5.76%      62.053ms     282.059us           220  
                       aten::_log_softmax_backward_data         5.33%      57.416ms         5.33%      57.416ms       5.742ms            10  
                                            aten::copy_         5.04%      54.314ms         5.04%      54.314ms     133.778us           406  
    autograd::engine::evaluate_function: AddmmBackward0         4.96%      53.432ms        22.40%     241.358ms       2.682ms            90  
```

```
(py38-cpu-perf-regression) nshulga@compile-st-c524xlarge-1:/fsx/users/nshulga/examples/word_language_model$ ATEN_CPU_CAPABILITY=avx2 python3 profile.py|head -n10
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                     aten::_log_softmax        29.88%     287.279ms        29.88%     287.279ms      28.728ms            10  
                                               aten::mm        17.55%     168.752ms        17.56%     168.810ms     937.833us           180  
                                            aten::addmm         6.72%      64.632ms        11.55%     111.066ms       1.234ms            90  
                       aten::_log_softmax_backward_data         6.57%      63.190ms         6.57%      63.190ms       6.319ms            10  
                                            aten::fill_         6.05%      58.173ms         6.05%      58.173ms     264.423us           220  
                                            aten::copy_         5.54%      53.288ms         5.54%      53.288ms     131.251us           406  
    autograd::engine::evaluate_function: AddmmBackward0         4.56%      43.840ms        24.00%     230.767ms       2.564ms            90  
```

**Comment by mruberry** [comment]
Discussion:

- let's file an issue to improve benchmarking -- possibly creating a benchmarking job

**Comment by imaginary-person** [comment]
@mruberry , there are some existing ATen op benchmarks at https://github.com/pytorch/pytorch/tree/master/benchmarks/operator_benchmark/pt, and we can run them with different `ATEN_CPU_CAPABILITY`. Some ops work better with AVX512, but there's significant run-to-run variation. I've noticed that preloading a third-party memory allocator (jemalloc/tcmalloc) produces more consistent results. I'll post results later.

I guess we can also benchmark on AWS m6i.64xlarge AWS instance, which is Ice Lake, and have 48 KB L1D caches. Thanks!

**Comment by imaginary-person** [comment]
Hi @malfet @ngimel @ezyang @VitalyFedyunin @mruberry, 

Even after we'd finish benchmarking & analyzing performance of ATen kernels (and potentially enable dispatch of some AVX512 kernels), can we still let AVX2 be the default CPU capability, so that users may use AVX512 ATen kernels by using the environment variable `ATEN_CPU_CAPABILITY=AVX512`?

There are at least 4 factors that would also have to be considered while benchmarking & comparing AVX2 & AVX512 performance of each ATen kernel:
1. Input(s) size(s) 
2. Number of threads
3. CPU platform
4. Memory allocator (AVX512 might do better with non-default memory allocators with large page-sizes.) 

For some combination of the aforementioned factors, some ATen kernels might exhibit better performance with AVX512, but worse with another combination.  However, currently, PyTorch dispatches an ATen kernel according to the highest CPU capability, so the dispatch doesn't consider these factors. So, would it be okay to add a Python API to change `ATEN_CPU_CAPABILITY` at runtime, which would allow users to fine-tune performance of workloads, based on knowledge gleaned via prior benchmarking? Thanks!

**Comment by malfet** [comment]
@imaginary-person I've already switched to `avx2` by default in https://github.com/pytorch/pytorch/pull/80253

As for the rest, I think majority of the users would not know which kernel to pick, so if we want to use AVX512, then we should code the heuristics into the AVX512 kernel implementation

**Comment by sanchitintel** [comment]
This specific issue was fixed in #85398. Thanks!
